query-engine/connectors/sql-query-connector/src/query_builder/select/mod.rs

```diff
@@ -1,89 +1,142 @@
-use std::{borrow::Cow, collections::BTreeMap};
+mod mysql;
+mod postgres;
+
+use std::borrow::Cow;
+
+use psl::datamodel_connector::Flavour;
 use tracing::Span;
 
 use crate::{
     context::Context,
-    filter::alias::{Alias, AliasMode},
-    model_extensions::{AsColumn, AsColumns, AsTable, ColumnIterator, RelationFieldExt},
+    filter::alias::Alias,
+    model_extensions::{AsColumns, AsTable, ColumnIterator, RelationFieldExt},
     ordering::OrderByBuilder,
     sql_trace::SqlTraceComment,
 };
 
 use quaint::prelude::*;
 use query_structure::*;
 
-pub const JSON_AGG_IDENT: &str = "__prisma_data__";
+use self::{mysql::MysqlSelectBuilder, postgres::PostgresSelectBuilder};
 
-#[derive(Debug, Default)]
-pub(crate) struct SelectBuilder {
-    alias: Alias,
-}
+pub(crate) const JSON_AGG_IDENT: &str = "__prisma_data__";
+
+pub(crate) struct SelectBuilder {}
 
 impl SelectBuilder {
-    pub(crate) fn next_alias(&mut self) -> Alias {
-        self.alias = self.alias.inc(AliasMode::Table);
-        self.alias
+    pub fn build(args: QueryArguments, selected_fields: &FieldSelection, ctx: &Context<'_>) -> Select<'static> {
+        match connector_flavour(&args) {
+            Flavour::Mysql => MysqlSelectBuilder::default().build(args, selected_fields, ctx),
+            Flavour::Postgres | Flavour::Cockroach => {
+                PostgresSelectBuilder::default().build(args, selected_fields, ctx)
+            }
+            _ => unreachable!("Connector does not support joined selects."),
+        }
     }
+}
 
-    pub(crate) fn build(
+pub(crate) trait JoinSelectBuilder {
```

Comment:
To deal with the differences of PG & MySQL (lateral join vs correlated sub-queries), I've went for a trait design with a bunch of unimplemented methods that handle the few differences that we have between both connectors.

It has grown a bit too much to my taste already with the addition of `_count` support, but I can't think of a cleaner way to avoid huge code duplication.